{"cells":[{"cell_type":"code","metadata":{},"source":["from dataclasses import dataclass\n\nimport pandas as pd\nimport altair as alt\n\n# Sequencing\n\n\n@dataclass\nclass Seq:\n    sample: str\n    at: int\n    data: str\n\n\n@dataclass\nclass ReadCountMatrix1:\n    @dataclass\n    class M:\n        sample: str\n        at: int\n\n    @dataclass\n    class D:\n        df: pd.DataFrame\n\n    m: M\n    d: D\n\n\n@dataclass\nclass ReadCountMatrix2:\n    @dataclass\n    class M:\n        sample1: str\n        sample2: str\n        at: int\n\n    @dataclass\n    class D:\n        df: pd.DataFrame\n\n    m: M\n    d: D\n\n\ndef load_local_reads(\n    seq: Seq,\n) -> ReadCountMatrix1.D:\n    df = ...\n    return pd.DataFrame(df=df)\n\n\ndef aggregate_reads(\n    rcm1: ReadCountMatrix1,\n    rcm2: ReadCountMatrix1,\n) -> ReadCountMatrix2.D:\n    df = ...\n    return pd.DataFrame(df=df)\n\n\n# Transfection and growth\n\n\n@dataclass\nclass Transfect:\n    sample: str\n    at: int\n    library: str\n\n\n@dataclass\nclass GrowthPhenotype:\n    @dataclass\n    class M:\n        sample: str\n        start: int\n        end: int\n\n    @dataclass\n    class D:\n        df: pd.DataFrame\n\n    m: M\n    d: D\n\n\ndef growth_phenotype(\n    t: Transfect,\n    rcm1: ReadCountMatrix1,\n    rcm2: ReadCountMatrix1,\n) -> GrowthPhenotype.D:\n    df = ...\n    return GrowthPhenotype.D(df=df)\n\n\n# Bulk RNA-seq\n\n\n@dataclass\nclass DifferentialGeneExpression:\n    @dataclass\n    class M:\n        sample1: str\n        sample2: str\n        at: int\n\n    @dataclass\n    class D:\n        df: pd.DataFrame\n\n    m: M\n    d: D\n\n\ndef combat_seq(\n    rcm: ReadCountMatrix2,\n) -> ReadCountMatrix2.D:\n    df = ...\n    return ReadCountMatrix2.D(df=df)\n\n\ndef deseq2(\n    rcm: ReadCountMatrix2,\n) -> DifferentialGeneExpression.D:\n    df = ...\n    return DifferentialGeneExpression.D(df=df)"],"id":null,"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{},"id":null,"attachments":null,"source":["# Load data"]},{"cell_type":"code","metadata":{},"source":["output_rcm_rcm_rcm1_seq = Seq(\n    at=1,\n    sample=\"healthy\",\n    data=\"healthy.fastq\"\n)","output_rcm_rcm_rcm1_seq"],"id":null,"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{},"source":["output_rcm_rcm_rcm2_seq = Seq(\n    at=1,\n    sample=\"tle\",\n    data=\"cri.fastq\"\n)","output_rcm_rcm_rcm2_seq"],"id":null,"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{},"id":null,"attachments":null,"source":["# Run analysis"]},{"cell_type":"code","metadata":{},"source":["output_rcm_rcm_rcm1 = ReadCountMatrix1(\n    m=ReadCountMatrix1.M(at=1, sample=\"healthy\"),\n    d=load_local_reads(\n        seq=output_rcm_rcm_rcm1_seq\n    )\n)","output_rcm_rcm_rcm1"],"id":null,"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{},"source":["output_rcm_rcm_rcm2 = ReadCountMatrix1(\n    m=ReadCountMatrix1.M(at=1, sample=\"tle\"),\n    d=load_local_reads(\n        seq=output_rcm_rcm_rcm2_seq\n    )\n)","output_rcm_rcm_rcm2"],"id":null,"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{},"source":["output_rcm_rcm = ReadCountMatrix2(\n    m=ReadCountMatrix2.M(sample1=\"healthy\", at=1, sample2=\"tle\"),\n    d=aggregate_reads(\n        rcm1=output_rcm_rcm_rcm1,\n        rcm2=output_rcm_rcm_rcm2\n    )\n)","output_rcm_rcm"],"id":null,"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{},"source":["output_rcm = ReadCountMatrix2(\n    m=ReadCountMatrix2.M(at=1, sample1=\"healthy\", sample2=\"tle\"),\n    d=combat_seq(\n        rcm=output_rcm_rcm\n    )\n)","output_rcm"],"id":null,"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{},"source":["output = DifferentialGeneExpression(\n    m=DifferentialGeneExpression.M(at=1, sample1=\"healthy\", sample2=\"tle\"),\n    d=deseq2(\n        rcm=output_rcm\n    )\n)","output"],"id":null,"execution_count":null,"outputs":[]}],"metadata":{},"nbformat":4,"nbformat_minor":0}